ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -174,6 +175,126 @@ func (r *RayClusterReconciler) rayClusterReconcile(ctx context.Context, request
+					r.Log.Info(fmt.Sprintf("The head Pod %s is already being deleted. Skip deleting this Pod.", headPod.Name))
+					continue
+				}
+				r.Log.Info(fmt.Sprintf(
+					"Delete the head Pod %s before the Redis cleanup. "+
+						"The storage namespace %s in Redis cannot be fully deleted if the GCS process on the head Pod is still writing to it.",
+					headPod.Name, headPod.Annotations[common.RayExternalStorageNSAnnotationKey]))
+				if err := r.Delete(ctx, &headPod); err != nil {
+					return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
+				}
+			}
+
+			// Delete all worker Pods if they exist.
+			for _, workerGroup := range instance.Spec.WorkerGroupSpecs {
+				workerPods := corev1.PodList{}
+				filterLabels = client.MatchingLabels{common.RayClusterLabelKey: instance.Name, common.RayNodeGroupLabelKey: workerGroup.GroupName}
+				if err := r.List(ctx, &workerPods, client.InNamespace(instance.Namespace), filterLabels); err != nil {
+					return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
+				}
+
+				for _, workerPod := range workerPods.Items {
+					if !workerPod.DeletionTimestamp.IsZero() {
+						r.Log.Info(fmt.Sprintf("The worker Pod %s is already being deleted. Skip deleting this Pod.", workerPod.Name))
+						continue
+					}
+					r.Log.Info(fmt.Sprintf(
+						"Delete the worker Pod %s. This step isn't necessary for initiating the Redis cleanup process.", workerPod.Name))
+					if err := r.Delete(ctx, &workerPod); err != nil {
+						return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
+					}
+				}
+			}
+
+			// If the number of head Pods is not 0, wait for it to be terminated before initiating the Redis cleanup process.
+			if len(headPods.Items) != 0 {
+				r.Log.Info(fmt.Sprintf(
+					"Wait for the head Pod %s to be terminated before initiating the Redis cleanup process. "+
+						"The storage namespace %s in Redis cannot be fully deleted if the GCS process on the head Pod is still writing to it.",
+					headPods.Items[0].Name, headPods.Items[0].Annotations[common.RayExternalStorageNSAnnotationKey]))
+				// Requeue after 10 seconds because it takes much longer than DefaultRequeueDuration (2 seconds) for the head Pod to be terminated.
+				return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+			}
+
+			// We can start the Redis cleanup process now because the head Pod has been terminated.
+			filterLabels = client.MatchingLabels{common.RayClusterLabelKey: instance.Name, common.RayNodeTypeLabelKey: string(rayv1alpha1.RedisCleanupNode)}
+			redisCleanupJobs := batchv1.JobList{}
+			if err := r.List(ctx, &redisCleanupJobs, client.InNamespace(instance.Namespace), filterLabels); err != nil {
+				return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
+			}
+
+			if len(redisCleanupJobs.Items) != 0 {
+				// Check whether the Redis cleanup Job has been completed.
+				redisCleanupJob := redisCleanupJobs.Items[0]
+				if redisCleanupJob.Status.Succeeded > 0 {
```

Comment:
Should we do something if the status is Failed? (Maybe log a message at least?)

Reply:
Updated https://github.com/ray-project/kuberay/pull/1412/commits/aca1b0a931f7ae1c56afc63097675af322b7cac9